pkg/network/bridge/plugin.go

```diff
@@ -189,53 +187,48 @@ func getAssignmentPath(bridge string, nic string) string {
 }
 
 func (nbdp *NetworkBridgeDevicePlugin) attachPods() {
-	pendingAssignments := list.New()
-
 	cli, err := dockerutils.NewClient()
 	if err != nil {
 		glog.V(3).Info("Failed to connect to Docker")
 		panic(err)
 	}
 
+	var attachMutex sync.Mutex
+
 	for {
-		select {
-		case assignment := <-nbdp.assignmentCh:
+		go func(assignment *Assignment) {
 			glog.V(3).Infof("Received a new assignment: %s", assignment)
-			pendingAssignments.PushBack(assignment)
-		default:
-			time.Sleep(time.Second)
-		}
-
-		for a := pendingAssignments.Front(); a != nil; a = a.Next() {
-			assignment := a.Value.(*Assignment)
-			glog.V(3).Infof("Handling pending assignment for: %s", assignment.DeviceID)
-
-			if time.Now().After(assignment.Created.Add(assignmentTimeout)) {
-				glog.V(3).Infof("Assignment for %s timed out", assignment.DeviceID)
-				pendingAssignments.Remove(a)
-				continue
-			}
-
-			containerID, err := cli.GetContainerIDByMountedDevice(assignment.ContainerPath)
-			if err != nil {
-				glog.V(3).Infof("Container was not found, due to: %s", err.Error())
-				continue
-			}
-
-			containerPid, err := cli.GetPidByContainerID(containerID)
-			if err != nil {
-				glog.V(3).Info("Failed to obtain container's pid, due to: %s", err.Error())
-				continue
+			for i := 1; i <= attachmentRetries; i++ {
+				if attached := func() bool {
+					containerID, err := cli.GetContainerIDByMountedDevice(assignment.ContainerPath)
+					if err != nil {
+						glog.V(3).Infof("Container was not found, due to: %s", err.Error())
+						return false
+					}
+
+					containerPid, err := cli.GetPidByContainerID(containerID)
+					if err != nil {
+						glog.V(3).Info("Failed to obtain container's pid, due to: %s", err.Error())
+						return false
+					}
+
+					attachMutex.Lock()
```

Comment:
why is the mutex needed? is it because of the netns.Set() call?

Reply:
I think so, I feel it would introduce some race problems without the lock.

Reply:
ok